### DISCUSSION TITLE
Table orders_status_history doesn't track history

### DISCUSSION THREAD

**Discussion by mc12345678** [zencart/zencart#6766] [General]
The data of table `orders_status_history` doesn't record actual historical information associated with the status but instead the status ID that was assigned to the order and the date/time recorded associated with that action. The following results in generation of debug logs in execution anywhere the status history is presented and where status text is changed causes incorrect user order record history modification. 

Reproduce by 
installing default store, 
add the demo data, 
Ready the store for a simple purchase as necessary
In the admin add an order status
Use the demo data user to place an order
In the admin for the order change the order status to the newly added status
On the catalog, access the user's order history to see the status. 
Now in the admin modify the text of the newly added status. 
On the catalog side, assuming the data is not cached, redisplay of the order status shows the new text. 
On the admin now delete that newly added status from being a selection.
**(EDIT ADDED START:)**
Note that deletion is prevented and a message is displayed.
Now in the admin, for the order, change the order status to any other order status.
On the catalog, access the user's order history to see the status is updated to have the "newly added" status and the additional other status. 
On the admin now delete that newly added status from being a selection and note it is not blocked.
**(EDIT ADDED END:)**
Reload the catalog side order history, see new debug log.
On the admin, load the order and also see that new debug log as well.

If attempt to create a new status to "restore" that text, issue persists for this order because the new status is given the next new number. Resolution for the debug log is to force create the "missing" record through a SQL query. 

Hope this customer presented information isn't expected in various localities to be more persistent (think on related order history table(s))

**Comment by drbyte** [comment]
Without seeing your (or recreating the situation in order to generate my own) debug log, it's not clear whether the issue recorded in the debug log is merely a null value due to invalid SQL join (of course due to the missing record) or a failed lookup of a displayable name (again due to the missing record), etc. There are other possibilities as well, depending on how fatal the error is that's reported.

We don't usually just randomly go creating missing records, because we don't know what should be in them. That said, we could provide a report and suggest ways to fix broken data.

Probably should at least:
- prevent deleting order_status records for which orders_status_history records exist
- have a fallback display string used if a missing record is detected, to avoid the debug log

**Reply by mc12345678** [reply]
On a ZC 1.5.8 system, where code in this area remains the same in 2.0.1 and forward, the below was the debug message received.  Store database shows had been upgraded from 1.5.5 more than a year ago, issue of non-existent orders_status_id applied to records created after that upgrade. Old records exist in the `orders_status_history` table where the orders_status_id is 0 and later records exist where both orders_status_id of 0 and 6 are both on a record and then more recently some records have had the orders_status_id of 6 applied, but neither status 0 nor status 6 currently exists.

Review of the store's order records having non-existent orders_status_ids do not appear to have any consistent condition accounting for how it came to have the non-existent orders_status_id and the particular orders_status_ids no longer exist (only that all such usages were not as the current/last usage). Generally speaking, in review of the recorded data, the value of 0 no longer existing doesn't surprise me as much as a missing non-negative value.

The area of the existing test is:
https://github.com/zencart/zencart/blob/0bbca88f71e3275af64ce6a78b44df234f63377c/admin/orders_status.php#L78-L89

That test only looks at the orders table, where the order will have only the most recent orders_status_id and is otherwise unrelated to the orders_status_history further allowing changes to the so-called history which is only really tracking a numeric designation (yes that number has an impact on operations such as software downloads, regardless the text information associated).
 
Sure, so my method of problem recreation wasn't quite complete having found this issue weeks ago and just finally deciding to try to write it out (Updated my original post with the additional step(s) to recreate): 
After creating and assigning this new status_id, update the record with some other status that is not about to be deleted.

Now, when deleting the newly created status_id, the "test" will not prevent deletion of the status record where doing that or editing the orders_status_name (especially after making customer notification) will also happen without delay.

````
[Today Now] Request URI: /admin/index.php?cmd=orders&oID=XXXX&action=edit, IP address: YYY.YYY.YYY.YYY, Language id 1
#0 /home/domain/admin/orders.php(936): zen_debug_error_handler()
#1 /home/domain/admin/index.php(11): require('/home/domain...')
--> PHP Warning: Undefined array key 6 in /home/domain/admin/orders.php on line 936.
````

**Reply by drbyte** [reply]
> Generally speaking, in review of the recorded data, the value of 0 no longer existing doesn't surprise me as much as a missing non-negative value.

`0` won't exist in `orders_status` because the `id` is an `auto_increment` field. 
Standard practice in relational database table design/operation is to "set foreign to zero on delete" (or null if applicable), when not deleting all cascading records. (ie: if you delete an orders_status record but don't also delete all the places where that id is used, such as orders_status_history, then set the id to 0 for those records in orders_status_history to indicate such).

That's the standard whether enforcing foreign-key relationships or not. But I'm not taking this conversation in the direction of establishing foreign-key schemas.

The point: `0` is expected in the `orders_status_history` table if no corresponding `orders_status` record exists for that entry.

>  Store database shows had been upgraded from 1.5.5 more than a year ago, issue of non-existent orders_status_id applied to records created after that upgrade. Old records exist in the orders_status_history table where the orders_status_id is 0 and later records exist where both orders_status_id of 0 and 6 are both on a record and then more recently some records have had the orders_status_id of 6 applied, but neither status 0 nor status 6 currently exists.
> Review of the store's order records having non-existent orders_status_ids do not appear to have any consistent condition accounting for how it came to have the non-existent orders_status_id and the particular orders_status_ids no longer exist (only that all such usages were not as the current/last usage). 

As far as encountering the orphaned id of `6` in your current dataset, that would seem to have been caused by manual deletion of the `6` record from `orders_status`. That could happen by direct database queries via something like phpMyAdmin, or perhaps a delete-and-restore-from-old of just that `orders_status` table, or from some 3rd party tool that the storeowner has used to manage their database outside of the Zen Cart Admin. No way to determine for certain without consulting whoever knows all the history of this store.

As to "fixing" the orphaned records for id=`6`, you could either manually insert said record into `orders_status` and specify `6` as the id (and maybe set a name like 'history unknown'), or you could renumber all uses of 6 to 0.

**Comment by scottcwilson** [comment]
> prevent deleting order_status records for which orders_status_history records exist

I could have sworn this was already done, but it's not.  Agreed, this is a good idea.

**Reply by mc12345678** [reply]
> > prevent deleting order_status records for which orders_status_history records exist
> 
> I could have sworn this was already done, but it's not. Agreed, this is a good idea.

Done for deleting a status that is currently applied against an order, but as described above is not stopped for one that is part of the order's history.

**Comment by drbyte** [comment]
So, I followed the steps @mc12345678 listed, and am blocked when trying to do the deletion in the Admin.
Turns out we already do block deletion of in-use orders_status records, whether renamed or not:
<img width="999" alt="Screen Shot 2024-09-22 at 11 58 23 AM" src="https://github.com/user-attachments/assets/dc7235e2-49fe-4091-a26b-f14c20e7b2ab">
<img width="1010" alt="Screen Shot 2024-09-22 at 12 02 17 PM" src="https://github.com/user-attachments/assets/705d1cd7-8677-445b-ad4d-2845a5f6423d">

**Reply by drbyte** [reply]
Here's the code, which does the requisite checks, regardless of the name of the orders_status records:
https://github.com/zencart/zencart/blob/959e2945c96a0c38290b99ef00f3446df0d06ff5/admin/orders_status.php#L75-L99

**Reply by drbyte** [reply]
Further digging in code shows that the status messages were added in v1.5.7a, but the blocking of deletions-if-in-use has been present since long before that.

**Reply by mc12345678** [reply]
> if-in-use

Narrow definition.
As implemented, in use means as a current value in the single record associated with each order and does not address other applicable use such as order history. 

Again, I note, the order history table only captures the numeric order status allowing any modification of the status designation. This then impacts display of all order records using that order status id including, if deleted, a blank description with associated debug log.

**Reply by drbyte** [reply]
> Again, I note, the order history table only captures the numeric order status allowing any modification of the status designation. This then impacts display of all order records using that order status id including, if deleted, a blank description with associated debug log.

Yes, that's by design. That's the nature of relational databases. Same exists with things like products vs products_description, ezpages vs ezpages_content, etc.

**Reply by drbyte** [reply]
> As implemented, in use means as a current value in the single record associated with each order and does not address other applicable use such as order history.

Line 78 above checks the `orders` table for any usages. Then line 86 checks whether that's non-zero and blocks the removal.
Subsequently, lines 90-97 block the removal if there's any usage of it in `orders_status_history`.

So, I'm not seeing where the problem is here. You keep saying it's used-but-not-checked in both `orders` and `orders_status_history` ... but the code quoted here clearly does check both.

**Comment by drbyte** [comment]
@mc12345678 Your described problem doesn't match the code.

Please post the debug logs you reported that you got.

Please also post the context in which this all was discovered, including which ZC version, which addons are affecting the code on the site where you discovered it (for example, is the Admin using some "other" way of deleting orders_status records, or some other non-core tools, etc etc etc).

**Reply by mc12345678** [reply]
````
[Today Now] Request URI: /admin/index.php?cmd=orders&oID=XXXX&action=edit, IP address: YYY.YYY.YYY.YYY, Language id 1
#0 /home/domain/admin/orders.php(936): zen_debug_error_handler()
#1 /home/domain/admin/index.php(11): require('/home/domain...')
--> PHP Warning: Undefined array key 6 in /home/domain/admin/orders.php on line 936.
````
I don't have the login information for the catalog side occurrence; however, the same code/condition occurs on the catalog side within the order history display and doesn't have any protections to prevent issues similar to above. I would expect the cause on the admin side to be easily found and similarly corrected on the catalog side to be corrected everywhere possible to occur not just where logs have been found.

**Comment by scottcwilson** [comment]
> Turns out we already do block deletion of in-use orders_status records, whether renamed or not:

LOL it helps if you test with a database that has more than zero orders! 

Thanks for verifying this.

**Reply by mc12345678** [reply]
> > Turns out we already do block deletion of in-use orders_status records, whether renamed or not:
> 
> LOL it helps if you test with a database that has more than zero orders!
> 
> Thanks for verifying this.

Yeah, though the original directions were effectively missing a step (add an additional status the order), one of the first things to do was to place an order. `Use the demo data user to place an order`.

As to "verifying" all that was verified was that an `orders_status_id` cannot be deleted if it is currently in the `orders` table and no prevention of deletion if the status has ever been used. 

This thread is not about the `orders` table but instead the `orders_status_history` table.

**Comment by drbyte** [comment]
That said, currently in the v210-alpha2 code, if the orders_status_history record is orphaned, we display a "blank" instead of a message of "Invalid".
<img width="1104" alt="Screen Shot 2024-09-26 at 2 31 06 PM" src="https://github.com/user-attachments/assets/aa02e2d1-4262-4920-be95-5de5c62177b8">

The code 'intended' to show "Invalid", but with the implementation of strict type-checking via `!==`, the `null` doesn't match `''` blank string.

orders.php line 1475 could be changed to use `!empty()` instead:
```diff
-<td class="dataTableContent text-right"><?php echo ($orders->fields['orders_status_name'] !== '' ? $orders->fields['orders_status_name'] : TEXT_INVALID_ORDER_STATUS); ?></td>
+<td class="dataTableContent text-right"><?php echo !empty($orders->fields['orders_status_name']) ? $orders->fields['orders_status_name'] : TEXT_INVALID_ORDER_STATUS; ?></td>
```
<img width="1098" alt="Screen Shot 2024-09-26 at 2 31 22 PM" src="https://github.com/user-attachments/assets/f7a2cfab-cc7b-40ba-b2be-f2969e54d525">

**Reply by drbyte** [reply]
Update: addressed in 1407a64b9
